---
title: "using QCkit-fastq package"
author: "Wenyue Xing"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style


## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

```{r}
knitr::opts_chunk$set(fig.width=6, fig.height=4,fig.align = "center") 


```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))


#I. Pre-processing of FASTQ files

```{R,include=FALSE}

library(devtools)
install_github("winnie0521/qckit",ref="cplusdocumentation")


```

```{r}
library(qckit)
infile <- system.file("extdata", "10^5_reads_test.fq.gz", package = "qckit")

```

#II. Metrics of FASTQ quality control


##1. dimension

*dimensions* function obtains the number of reads and number of positions from a seqTools processed FASTQ file. The sample file has 25,000 reads and 100 positions. 

```{r}

fseq <- seqTools::fastqq(infile)
nc <- dimensions(fseq,"positions")
nc

nr <- dimensions(fseq,"reads")
nr
```

*sequence_length* function generates distribution plot of the length of all reads. The generated plot would show the sequence length of all the sequences throughout the file. The plot is considered an indication of good data quality is all sequences have the same sequence length with no deviations. The following plot shows that all reads in the file have sequence length of 100. 

```{r}
plot_sequence_length(fseq,writefile=FALSE)


```

##2.  quality score per sequence statistics

*basic_statistics* function calculates the quality score per sequence statistics including mean, median and quantiles. We will use the result to create a quality score distribution per position plot to present the overall dispersion and concentration of quality score per position of the FASTQ file. 

```{r}

bs <- basic_stat(infile,FALSE)
knitr::kable(head(bs))
plot_quality_score(bs,writefile=FALSE)
```

##3. nucleotide sequence content

*sequence_content* calculates the total number of each nucleotide sequence content per sequence throughout the file. 
```{r}


scA <- sequence_content(fseq, content = "A")
scA


```

*plot_sequence_conent* use the *sequence_content* function and plot the percentage of all nucleotide sequence content per position. The plot would be considered an indication of good data quality when the percentage of each nucleotide sequence content is not significantly larger than other nucleotide sequence contents. 

```{r}

plotSeqContent(fseq,nr,nc)

```


##4. GC content


Function *GC_content* utilizes the result from the C++_calling function *GC_per_read* and produce the data frame that is more suitable for the plotting purpose as well as for saving to data file for later access. 

```{r}

gc_df <- GC_content(infile)

knitr::kable(head(gc_df))


```


The *GC_content_plot* function takes the input from previous function and generates plot of distribution of GC nucleotide sequence content percentage. The graph would be an indication of good data quality when the  GC percentage concentrated around 30 to 50. 


```{r}

GC_content_plot(nc,gc_df)


```


##5. Per read sequence quality score

*plot_perseq_quality* function extracts the mean quality score per read and generate a histogram of this statistics.It takes the input of the path to the gzipped FASTQ files. It also enables plot saving if set *writefile* to TRUE and input a *prefix* value to customize name of saved file. The histogram would be considered an indication of good data quality when the majority of reads have high quality score(around 30). If a significant portion of read are with low quality score, say less than 20, then data is problematic. 


```{r}

plot_perseq_quality(infile,writefile=FALSE)

```

##6. Kmers & Overrepresented Kmers

*kmer_count* function produces the per position kmer count with given path to the FASTQ file and the kmer length specified. 

```{r}

km <- Kmer_count(infile,k=6)
knitr::kable(km[1:20,1:10])

```


The function generates list of overrepresented kmers with its maximum obs/exp ratio and the position of maximum obs/exp ratio. 


```{r}

overkm <-qckit::overrep_kmer(infile,7,nc,nr)
knitr::kable(head(overkm,n=10))


```

##6. Overrepresented Sequence

Taking in the output generated from *cal_over_rep_seq*, function *overrepresented_sequence* produce the overrepresented sequence table by selecting the unique sequences that have counts larger than 0.1% of the total reads of the data file. The results would be displayed as table with decreasing order of counts. 


```{r}

overrep_seq <- qckit::overrepresented_sequence(infile,nr,"test")
knitr::kable(head(overrep_seq,n = 5))

```

Function *overrepresented_plot* would take the overrepresented sequence table as input and present a density plot of the counts and mark the top 5 overrepresented sequence in red. 


```{r}


overrep_plot(overrep_seq,FALSE)


```

